Foundation/URLSession/NativeProtocol.swift

```diff
@@ -0,0 +1,715 @@
+// Foundation/NSURLSession/_NativeProtocol.swift - NSURLSession & libcurl
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2016 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+// -----------------------------------------------------------------------------
+///
+/// This file has the common implementation of Native protocols like HTTP,FTP,Data 
+/// These are libcurl helpers for the URLSession API code.
+/// - SeeAlso: https://curl.haxx.se/libcurl/c/
+/// - SeeAlso: NSURLSession.swift
+///
+// -----------------------------------------------------------------------------
+
+import CoreFoundation
+import Dispatch
+
+internal let enableLibcurlDebugOutput: Bool = {
+    return  (ProcessInfo.processInfo.environment["URLSessionDebugLibcurl"] != nil)
+}()
+internal let enableDebugOutput: Bool = {
+    return (ProcessInfo.processInfo.environment["URLSessionDebug"] != nil)
+}()
+
+class _NativeProtocol: URLProtocol, _EasyHandleDelegate {
+    internal var easyHandle: _EasyHandle!
+    internal var totalDownloaded = 0
+    internal lazy var tempFileURL: URL = {
+        let fileName = NSTemporaryDirectory() + NSUUID().uuidString + ".tmp"
```

Comment:
- don't we have a better/safer way to create temp files somewhere
 - is `NSTemporaryDirectory()` guaranteed to end in a `/` ?

Reply:
This change has been incorporated into the code base through https://github.com/apple/swift-corelibs-foundation/pull/1195.Hence I have just refactored and pulled it to NativeProtocol as it can be used by multiple protocols like FTP and HTTP